feature-terminate-insurance/src/main/kotlin/com/hedvig/android/feature/terminateinsurance/navigation/TerminateInsuranceGraph.kt

```diff
@@ -1,47 +1,101 @@
+    enterTransition = { MotionDefaults.sharedXAxisEnter(density) },
+    exitTransition = { MotionDefaults.sharedXAxisExit(density) },
+    popEnterTransition = { MotionDefaults.sharedXAxisPopEnter(density) },
+    popExitTransition = { MotionDefaults.sharedXAxisPopExit(density) },
   ) {
-    animatedComposable<TerminateInsuranceDestinations.TerminationDate> { navBackStackEntry ->
-      val viewModel: TerminateInsuranceViewModel = koinViewModel(viewModelStoreOwner = navBackStackEntry) {
-        parametersOf(insuranceId)
+    animatedComposable<TerminateInsuranceDestination.StartStep> {
+      val viewModel: TerminationStartStepViewModel = koinViewModel { parametersOf(insuranceId) }
+      TerminationStartDestination(
+        viewModel = viewModel,
+        retryLoad = { viewModel.retryToStartTerminationFlow() },
+        navigateToNextStep = { terminationStep ->
+          viewModel.handledNextStepNavigation()
+          navController.navigate(terminationStep.toTerminateInsuranceDestination())
+        },
+      )
+    }
+    animatedComposable<TerminateInsuranceDestination.TerminationDate> {
+      val shouldFinishFlowOnBack = run {
+        val previousBackstackEntryRoute = navController.previousBackStackEntry?.destination?.route
+        previousBackstackEntryRoute == createRoutePattern<TerminateInsuranceDestination.StartStep>()
       }
-      TerminationStepDestination(
+      BackHandler(shouldFinishFlowOnBack) {
+        finishTerminationFlow()
+      }
+      val viewModel: TerminationDateViewModel = koinViewModel { parametersOf(this.minDate, this.maxDate) }
+      TerminationDateDestination(
         viewModel = viewModel,
         windowSizeClass = windowSizeClass,
-        navigateToSuccessScreen = {
-          navController.navigate(
-            TerminateInsuranceDestinations.TerminationSuccess,
-            navOptions {
-              popUpTo<Destinations.TerminateInsurance>()
-            },
-          )
+        navigateToNextStep = { terminationStep ->
+          viewModel.handledNextStepNavigation()
+          navController.navigate(terminationStep.toTerminateInsuranceDestination())
+        },
+        navigateBack = {
+          if (shouldFinishFlowOnBack) {
+            finishTerminationFlow()
+          } else {
+            navController.navigateUp() || navigateUp()
+          }
         },
-        navigateBack = { navController.navigateUp() || navigateUp() },
-        finishTerminationFlow = finishTerminationFlow,
+      )
+    }
+    animatedComposable<TerminateInsuranceDestination.TerminationSuccess> {
+      BackHandler { finishTerminationFlow() }
+      TerminationSuccessDestination(
+        terminationDate = this.terminationDate,
+        surveyUrl = this.surveyUrl,
+        windowSizeClass = windowSizeClass,
+        navigateBack = finishTerminationFlow,
+      )
+    }
+    animatedComposable<TerminateInsuranceDestination.TerminationFailure> {
+      BackHandler { finishTerminationFlow() }
```

Comment:
Should we make this go back to the previous step? In case you would want to retry

Reply:
Took this to Slack as I don't know myself either. If backend supports this then we will make it so.